### ISSUE TITLE
[Question]infinite threshold in gbdt tree trained with trainset containing nan x data but well-bounded for the rest

### ISSUE THREAD

**Issue by HelloHumanWorld** [microsoft/LightGBM#5894]
I trained a gbdt model. y data is all data with values between -1 and 1, no NAN. x data has values between -3 and 3, but with some NAN values. I dump model and check the jason format of the model. I found that there was infinite threshold in some of the trees in the model. Any idea how this happened? 

Thank you.

**Comment by jameslamb** [comment]
Thanks for using LightGBM.

Please provide a [minimum, reproducible example](https://stackoverflow.com/help/minimal-reproducible-example)... code we could run to reproduce the issue you're seeing, as well as the details necessary to run it (which were asked for in the issue template), like:

* version of LightGBM
* how you installed LightGBM
* how you are using LightGBM (Python package? R package? CLI?)

Without that information, it's very unlikely that you'll get an answer here.

**Comment by HelloHumanWorld** [comment]
> Thanks for using LightGBM.
> 
> Please provide a [minimum, reproducible example](https://stackoverflow.com/help/minimal-reproducible-example)... code we could run to reproduce the issue you're seeing, as well as the details necessary to run it (which were asked for in the issue template), like:
> 
> * version of LightGBM
> * how you installed LightGBM
> * how you are using LightGBM (Python package? R package? CLI?)
> 
> Without that information, it's very unlikely that you'll get an answer here.

Dataset is huge. Not sure if I can create an example with a small dataset. Using python package installed by pip

**Comment by HelloHumanWorld** [comment]
The dataset is huge. Let me figure out how to create an example by a small set. I am using lgb python package's scikit-learn interface with version 3.3.5.

Actually, I dumped the booster by dump_model() to a jason format after the model is trained. And try to recreate the splitting logics by the jason format. 

My guessing is that if x values have Nan, when algo tries to split, it will also check whether inf value is a proper splitting-point in each step, and in the case that algo decides inf value as a splitting-point, it actually splits data into two groups(group 1 is all non-nan data, group2 is all nan data) in such way that algo finds it minimizing the loss at that step.

**Comment by guolinke** [comment]
@HelloHumanWorld it may due to the missing value handling. The NaNs are treated as missing values. For the node with inf threshold, it actually means the split of non-missing values (<= inf) and missing values (> inf).

**Comment by HelloHumanWorld** [comment]
> @HelloHumanWorld it may due to the missing value handling. The NaNs are treated as missing values. For the node with inf threshold, it actually means the split of non-missing values (<= inf) and missing values (> inf).

Thank you for the reply. I guess the same way. However, I have to point out that this handling is inconsistent with the document. As I remember, the document says that the algo decides splitting based on the non-missing values, and followed by assigning the missing values to one of two sides by mimizing loss.

**Comment by guolinke** [comment]
> > @HelloHumanWorld it may due to the missing value handling. The NaNs are treated as missing values. For the node with inf threshold, it actually means the split of non-missing values (<= inf) and missing values (> inf).
> 
> Thank you for the reply. I guess the same way. However, I have to point out that this handling is inconsistent with the document. As I remember, the document says that the algo decides splitting based on the non-missing values, and followed by assigning the missing values to one of two sides by mimizing loss.

Thank you for pointing out. Yeah, it actually does this in one-step (considering both non-missing and missing values at the same time), not the two-step.
